katsdpcontroller/generator.py

```diff
@@ -1642,30 +1644,45 @@ def _sky_model_url(data_url, continuum_name, target):
     return url
 
 
-def _normalise_target_name(name, used):
-    name = re.sub(r'[^-A-Za-z0-9_]', '_', name)
-    if name not in used:
+class TargetMapper:
+    """Assign normalised names to targets.
+
+    Each target passed to :meth:`__call__` is given a unique name containing
+    only alphanumeric characters, dashes and underscores. Passing the same
+    target will return the same name.
+
+    This class is not thread-safe.
+    """
+
+    def __init__(self) -> None:
+        self._cache: Dict[katpoint.Target, str] = {}
+        self._used: Set[str] = set()
+
+    def __call__(self, target: katpoint.Target) -> str:
+        if target in self._cache:
+            return self._cache[target]
+        name = re.sub(r'[^-A-Za-z0-9_]', '_', target.name)
+        if name in self._used:
+            i = 1
+            while '{}_{}'.format(name, i) in self._used:
+                i += 1
+            name = '{}_{}'.format(name, i)
+        self._cache[target] = name
+        self._used.add(name)
```

Comment:
Not that it matters much since, as you say, this is not thread safe anyway, but wouldn't reserving the name before adding it to the cache make more sense.

Reply:
I don't follow - what change do you want to make?

Reply:
Nothing important, just swapping the order of self._cache... and self._used...